### PR TITLE
groupByTags Performance improvements + fix setting consolidator per group

### PR DESCRIPTION
**Pull request by shanson7**
Use pre-parsed tag for name, only get the summarizeCons once.

Benchmarks for just this set of changes
```
benchmark                                              old ns/op      new ns/op      delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      11799          10548          -10.60%
BenchmarkGroupByTags10Series_100Datapoints_half-8      15975          14753          -7.65%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      220073         217893         -0.99%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      295150         301724         +2.23%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     14854710       13961473       -6.01%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     305783470      16255546       -94.68%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     1739274046     1734320020     -0.28%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     685210547      411100140      -40.00%

benchmark                                              old MB/s     new MB/s     speedup
BenchmarkGroupByTags10Series_100Datapoints_full-8      1016.96      1137.60      1.12x
BenchmarkGroupByTags10Series_100Datapoints_half-8      751.16       813.37       1.08x
BenchmarkGroupByTags10Series_10kDatapoints_full-8      5452.73      5507.28      1.01x
BenchmarkGroupByTags10Series_10kDatapoints_half-8      4065.72      3977.13      0.98x
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     807.82       859.51       1.06x
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     39.24        738.21       18.81x
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     689.94       691.91       1.00x
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     1751.29      2919.00      1.67x

benchmark                                              old allocs     new allocs     delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      40             28             -30.00%
BenchmarkGroupByTags10Series_100Datapoints_half-8      64             52             -18.75%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      40             28             -30.00%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      64             52             -18.75%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     20056          10054          -49.87%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     52289          41880          -19.91%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     22037          12035          -45.39%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     54386          42968          -20.99%

benchmark                                              old bytes     new bytes     delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      5242          4761          -9.18%
BenchmarkGroupByTags10Series_100Datapoints_half-8      7310          6826          -6.62%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      6478          5870          -9.39%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      15868         13792         -13.08%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     5110153       4790036       -6.26%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     9540243       9124264       -4.36%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     5522092       5202708       -5.78%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     207434013     140563990     -32.24%
```

Combined with #1164 and #1158 :

```
benchmark                                              old ns/op      new ns/op     delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      11799          10599         -10.17%
BenchmarkGroupByTags10Series_100Datapoints_half-8      15975          14098         -11.75%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      220073         184092        -16.35%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      295150         226125        -23.39%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     14854710       9008327       -39.36%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     305783470      15733034      -94.85%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     1739274046     212949763     -87.76%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     685210547      325118333     -52.55%

benchmark                                              old MB/s     new MB/s     speedup
BenchmarkGroupByTags10Series_100Datapoints_full-8      1016.96      1132.15      1.11x
BenchmarkGroupByTags10Series_100Datapoints_half-8      751.16       851.15       1.13x
BenchmarkGroupByTags10Series_10kDatapoints_full-8      5452.73      6518.46      1.20x
BenchmarkGroupByTags10Series_10kDatapoints_half-8      4065.72      5306.78      1.31x
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     807.82       1332.10      1.65x
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     39.24        762.73       19.44x
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     689.94       5635.13      8.17x
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     1751.29      3690.96      2.11x

benchmark                                              old allocs     new allocs     delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      40             28             -30.00%
BenchmarkGroupByTags10Series_100Datapoints_half-8      64             52             -18.75%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      40             28             -30.00%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      64             52             -18.75%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     20056          10044          -49.92%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     52289          41968          -19.74%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     22037          10234          -53.56%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     54386          41832          -23.08%

benchmark                                              old bytes     new bytes     delta
BenchmarkGroupByTags10Series_100Datapoints_full-8      5242          4761          -9.18%
BenchmarkGroupByTags10Series_100Datapoints_half-8      7310          6825          -6.63%
BenchmarkGroupByTags10Series_10kDatapoints_full-8      6478          5838          -9.88%
BenchmarkGroupByTags10Series_10kDatapoints_half-8      15868         13513         -14.84%
BenchmarkGroupByTags10kSeries_100Datapoints_full-8     5110153       4788323       -6.30%
BenchmarkGroupByTags10kSeries_100Datapoints_half-8     9540243       9205551       -3.51%
BenchmarkGroupByTags10kSeries_10kDatapoints_full-8     5522092       4827670       -12.58%
BenchmarkGroupByTags10kSeries_10kDatapoints_half-8     207434013     87353387      -57.89%
```

### CODE REVIEW

**Comment by Dieterbe** [comment]
the `Clean up tests/benchmarks` commit is confusing.

before we tested every combo of numseries 1/10/100/1000  and input series generation via 2x test.RandFloats10k / test.RandFloats10k,test.RandFloatsWithNulls10k  / test.RandFloatsWithNulls10k,test.RandFloatsWithNulls10k

why do we now benchmark fewer amounts of series and none of the RandFloatsWithNulls variants? i suppose the reason for the latter is because we don't want to conflate performance of the aggregator with that of GBT, but if that's the case, why do we still test different numbers of datapoints per series? I suppose we shouldn't benchmark every combo of every possible parameter as that makes the output  unwieldy and the runtime longer.
different numbers of points per series could be useful to see the effect of the data memory use on that of the metadata stuff that GBT is doing.
But if anything the number of input series seems like a parameter we should explore more, like we used to

File: expr/func_groupbytags_test.go [hunk 1]
```diff
@@ -257,64 +257,66 @@ func testGroupByTags(name string, in []models.Series, out []models.Series, agg s
...
-func benchmarkGroupByTags(b *testing.B, numSeries int, fn0, fn1 func() []schema.Point) {
+func benchmarkGroupByTags(b *testing.B, numSeries, numGroups int, fn0, fn1 func() []schema.Point) {
 	var input []models.Series
 	tagValues := []string{"tag1", "tag2", "tag3", "tag4"}
 	for i := 0; i < numSeries; i++ {
-		tags := make(map[string]string, len(tagValues))
-
-		for t, tag := range tagValues {
-			tags[tag] = strconv.Itoa(t)
-		}
```
[the remark this answers is not in the document]
Comment (shanson7) [reply expr/func_groupbytags_test.go]: It was "used" in that it was populated. It was never used after that though. Enough to satisfy the compiler!

**Comment by shanson7** [comment]
Yeah, at this point I was still trying to isolate the performance issues I was seeing with the aggregation. I think the two variables relevant to GBT are number of input series and output series. Maybe whether or not to include 'name'.

File: expr/func_groupbytags.go [hunk 2]
```diff
@@ -59,7 +62,7 @@ func (s *FuncGroupByTags) Exec(cache map[Req][]models.Series) ([]models.Series,
 	if !useName {
 		// if all series have the same name, name becomes one of our tags
 		for _, serie := range series {
-			thisName := strings.Split(serie.Target, ";")[0]
+			thisName := serie.Tags["name"]
```
Comment (Dieterbe) [inline expr/func_groupbytags.go]: right on.
it is probably not clear, but it looks like we do try to assure the name tag is always set for any series (FuncGet sets it initially and any other function that creates new series should set it as well I think)

File: expr/func_groupbytags.go [hunk 3]
```diff
@@ -105,8 +106,9 @@ func (s *FuncGroupByTags) Exec(cache map[Req][]models.Series) ([]models.Series,
 	aggFunc := getCrossSeriesAggFunc(s.aggregator)
 
 	// Now, for each key perform the requested aggregation
+	cons, queryCons := summarizeCons(series)
+
 	for name, groupSeries := range groups {
-		cons, queryCons := summarizeCons(series)
```
Comment (Dieterbe) [inline expr/func_groupbytags.go]: this seems incorrect. it's possible for input series to have different consolidators defined, or am I missing something?

**Comment by shanson7** [comment]
I rebased to be (I believe) what you were asking for, though the commit order shown here is different than I get with `git log`

File: expr/func_groupbytags.go (hunk 3 above)
Comment (shanson7) [reply expr/func_groupbytags.go]: That's true, but I think that's true with many of the aggregation functions and it seems first come first serve. 

In general, I think grouping here would be *very* awkward with mixing consolidators and the user is likely doing something...odd.

**Comment by shanson7** [comment]
Regarding the benchmarks, what do you think of:

input series: 1, 10, 100, 10k, 100k
output series: 1, same as input, then if applicable: 10, 100, 10k

This would result in 15 benchmarks.

**Comment by Dieterbe** [comment]
yes, but let's also include 1k.

**Comment by shanson7** [comment]
For both input and output? That would bump the number of benchmarks to 21. Should be fine

**Comment by Dieterbe** [comment]
yeah, i guess.

File: expr/func_groupbytags.go (hunk 3 above)
Comment (shanson7) [reply expr/func_groupbytags.go]: I think this is more correct, though in the case that the groups are mixed, there is still an arbitrary consolidator chosen. But when 'name' is one of the groupByTags, I could this working better

**Pull request merged**